novelwriter/gui/doceditor.py

```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
         else:
             text = self._trMainCount.format("0", "+0")
         self.wordsText.setText(text)
+
+    def updateVimModeStatusBar(self, modeName: str) -> None:
+        """Update the vim Mode status information."""
+        self.vimStatus.setText(self.tr(modeName))
+
+
+class VimState:
+    """Minimal Vim state machine."""
+
+    __slots__ = (
+        "PREFIX_KEYS",
+        "SUFFIX_KEYS",
+        "VISUAL_PREFIX_KEYS",
+        "_internalClipboard",
+        "_mode",
+        "_normalCommand",
+        "_visualCommand",
+    )
+
+    def __init__(self) -> None:
+        self.PREFIX_KEYS = ["d", "y", "g", "z"]
+        self.VISUAL_PREFIX_KEYS = ["g"]
+        self.SUFFIX_KEYS = ["w"]
+        self._mode: nwVimMode = nwVimMode.NORMAL
+        self._normalCommand: str = ""
+        self._visualCommand: str = ""
+        self._internalClipboard = ""
+        self.setMode(nwVimMode.NORMAL)
+
+    def resetCommand(self) -> None:
+        """Reset internal vim command."""
+        self._normalCommand = ""
+        self._visualCommand = ""
+
+    def setMode(self, new_mode: nwVimMode) -> None:
+        """Switch vim mode."""
+        self._mode = new_mode
+        self.resetCommand()
+
+    def getMode(self) -> nwVimMode:
+        """Return current vim mode."""
+        return self._mode
```

Comment:
Generally, I try to avoid get methods if they are plain and simple like this one and instead use Python property instead.

```suggestion
    @property
    def mode(self) -> nwVimMode:
        """Return current vim mode."""
        return self._mode
```